api/sip/client.py

```diff
@@ -252,127 +248,82 @@ def __init__(self, target_server, target_port, login_user_id=None,
             escaped = self.separator
         self.separator_re = re.compile(escaped + "([A-Z][A-Z])")
 
+        self.sequence_number = 0
+        self.connection = None
         self.login_user_id = login_user_id
         self.login_password = login_password
         if login_user_id and login_password:
             # We need to log in before using this server.
-            self.logged_in = False
             self.must_log_in = True
         else:
             # We're implicitly logged in.
-            self.logged_in = True
             self.must_log_in = False
 
-        # socket_lock controls access to the socket connection to the
-        # SIP2 server.
-        #
-        # We need to use an RLock here because both connect() and
-        # make_request() require the lock, and make_request() will end
-        # up calling connect() if there's an error.
-        self.socket_lock = threading.RLock()
-
-        # The only reason connect would be false is that we're running
-        # a unit test and don't actually want to use a server.
-        if connect:
-            self.connect()
-
-    def login(self, *args, **kwargs):
-        """Log in to the SIP server."""
-        return self.make_request(
-            self.login_message, self.login_response_parser,
-            *args, **kwargs
-        )
+    def login(self):
+        """Log in to the SIP server if required."""
+        if self.must_log_in:
+            response = self.make_request(
+                self.login_message, self.login_response_parser,
+                self.login_user_id, self.login_password, self.location_code
+            )
+            if response['login_ok'] != '1':
+                raise IOError("Error logging in: %r" % response)
+            return response
 
     def patron_information(self, *args, **kwargs):
-        """Get information about a patron, possibly also verifying their
-        password.
+        """Get information about a patron.
         """
         return self.make_request(
             self.patron_information_request, self.patron_information_parser,
             *args, **kwargs
         )
 
+    def end_session(self, *args, **kwargs):
+        """Send end session message."""
+        return self.make_request(
+            self.end_session_message, self.end_session_response_parser,
+            *args, **kwargs
+        )
+
     def connect(self):
         """Create a socket connection to a SIP server."""
-        with self.socket_lock:
-            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            try:
-                sock.connect((self.target_server, self.target_port))
-            except TypeError:
-                raise IOError(
-                    "Could not connect to %s:%s" % (
-                        self.target_server, self.target_port
-                    )
+        try:
```

Comment:
One last thing. In this method, check if there's an active connection already, and call `disconnect` if necessary. That way if there's any bad code left in the system it won't be able to create dangling connections.

Reply:
Good call. I think the garbage collector should handle this since connection goes out of scope here and a new connection is created. Good to look both ways and be explicit though. Added that in.